openedx/features/content_tagging/models.py

```diff
@@ -0,0 +1,106 @@
+"""
+Content Tagging models
+"""
+from django.db import models
+from model_utils.managers import InheritanceManager
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import CourseKey, UsageKey
+from openedx_tagging.core.tagging.models import Taxonomy
+from organizations.models import Organization
+
+
+class ContentTaxonomyManager(InheritanceManager):
+    """
+    Manages ContentTaxonomy objects, providing custom utility methods.
+
+    Inherits from InheritanceManager so that subclasses of ContentTaxonomy can be selected in a queryset.
+    """
+
+    def filter_enabled(self, org: Organization = None, enabled=True) -> models.QuerySet:
+        """
+        Returns a query set filtered to return the ContentTaxonomy objects enabled for the given org.
+        """
+        queryset = self
+        if enabled is not None:
+            queryset = queryset.filter(enabled=enabled)
+            org_filter = models.Q(org_owners=None)
+            if org:
+                org_filter |= models.Q(org_owners=org)
+            queryset = queryset.filter(org_filter)
+        return queryset
+
+
+class ContentTaxonomy(Taxonomy):
```

Comment:
I'm not convinced that it makes sense to make this a subclass of `Taxonomy`. Because it's not a true subclass, right? When `ObjectTag` [calls](https://github.com/openedx/openedx-learning/blob/821a53c38da47addec3135d32ac1ce70544c5263/openedx_tagging/core/tagging/models.py#L405) `self.taxonomy.validate_object_tag(self)`, there's no virtual dispatch involved. Regardless of whether there's a `ContentTaxonomy` or not, the `self.taxonomy` property always returns a `Taxonomy` instance, and only the base class's `validate_object_tag()` will be called. You would need to use something like [`django-polymorphic`](https://django-polymorphic.readthedocs.io/en/stable/) to get this to work (which is not what I am recommending; it's too much magic and we have enough dependencies already).

Reply:
Oh bummer. You're absolutely right. I had fooled myself into thinking this was ok in my tests, because when I created the invalid tags, their taxonomies are ContentTaxonomy instances. But they're not after they're refetched from the database :(

I guess I have to rethink this whole architecture then. Blech.

Reply:
Oh, I think we can have the same problem with [system-defined taxonomies](https://docs.google.com/document/d/1k2-ziCkCGT8AcGh2zorVinF0MH-2bF6_mAl0SFxr7EQ/edit#heading=h.8hqkzapyty1f)

Reply:
Is there a way to check if a model is an instance of ContentTaxonomy? (with `isistance` maybe?)

I remember the `validate_object_tag` function, it was first thought to be inside the API. We could check and depending of the instance, to call one and other function?

Reply:
The exact same thing has happened to me before... sorry I didn't spot this sooner. 

I didn't suggest any solutions because I didn't want to bias you one way or another, but I think there are still some nice ways to achieve a similar architecture without too many changes. One thing to think about for example would be if there is only one database/django model for `Taxonomy` but it has a field called `type` and that type determines the runtime behavior, and loads a different class or plugin to handle the implementation details. Basically, put your subclasses and hierarchy into pure python classes and keep the SQL data model simpler or even the same for all of them. But there are other ways to go too, that's just one idea.

Reply:
Thanks @bradenmacdonald  :)

I'm working through some options, and since users of the tagging (and content_tagging) modules should always go through the APIs, we can make sure they always return the right classes, and limit/discourage access to models via any other methods.

>  One thing to think about for example would be if there is only one database/django model for Taxonomy but it has a field called type and that type determines the runtime behavior, and loads a different class or plugin to handle the implementation details. Basically, put your subclasses and hierarchy into pure python classes and keep the SQL data model simpler or even the same for all of them.

👍 , and I hear that you don't want to introduce any "magic".. but [edx-platform already imports `django-model-utils`](https://github.com/openedx/edx-platform/blob/7ef20f0b919cdd3a7be42822bd6f375cfb0439f1/requirements/edx/base.in#L55), and its [InheritanceManager](https://django-model-utils.readthedocs.io/en/latest/managers.html?highlight=InheritanceManager#inheritancemanager) is made for this kind of thing. I haven't figured out how it works -- using it also doesn't change the migrations, and it doesn't add to the query count -- which puts it firmly in the "magic" camp, but..

Is using really any different from doing my own casting/creation/plugin management with a `Taxonomy.type` field?

Reply:
I've tried this and it's working well.. what do you think?

See https://github.com/openedx/openedx-learning/pull/60 + [4f297f6](https://github.com/openedx/edx-platform/pull/32518/commits/4f297f67ba2908874c7101be9f3fd577dc5ea2b9), specifically:

* [`oel_tagging.api` changes](https://github.com/openedx/openedx-learning/pull/60/files#diff-951a5c4d32e1301ac1c30419ebce4ada8ae17db6c93f4c2a5a34caac9eb30e0a), which were clean and minimal.
* [`oel_tagging.get_taxonomies` test](https://github.com/openedx/openedx-learning/pull/60/files#diff-6cda6f74e9f6bf053923d015688ea76e50cfec92ef244df48aaa3942288fcbb5R34-R39), which verifies that only a single query is used even if we're fetching multiple types of Taxonomies.
* [`ContentTaxonomy.validate_object_tag` test](https://github.com/openedx/edx-platform/pull/32518/commits/4f297f67ba2908874c7101be9f3fd577dc5ea2b9#diff-ca73cded4994ab137e40fc7011cee435f10feaa439af275937a282c5a8acb5b0R215-R232) that verifies the object tag validation.

Reply:
@pomegranited that seems fine with me in principle, as long as the code and APIs are carefully constructed to make sure `select_subclasses` is used. But this is really more of an @ormsbee question so I'd like to hear his thoughts.

> I haven't figured out how it works -- using it also doesn't change the migrations, and it doesn't add to the query count

It still uses only one query, but that query does an additional `JOIN` per registered subclass. So whether or not this is acceptable may depend on how many subclasses there are. If it's 1-2, should be fine. But if there are many, the query performance will be slower, even though it's still technically a single query. Are any of these `select_subclasses` on a hot path that will be called frequently, for example when retrieving tags? Or just for things like editing taxonomies?

Reply:
@bradenmacdonald @pomegranited 

> It still uses only one query, but that query does an additional JOIN per registered subclass. So whether or not this is acceptable may depend on how many subclasses there are. If it's 1-2, should be fine. But if there are many, the query performance will be slower, even though it's still technically a single query

Adding more context with this in mind. For [system-defined taxonomies](https://docs.google.com/document/d/1k2-ziCkCGT8AcGh2zorVinF0MH-2bF6_mAl0SFxr7EQ/edit#heading=h.squk6ryb7sqc), we were thinking of creating a subclass for each taxonomy, since each one had its special way of validating (`validate_object_tag`). With the current approach, in the short term it would already be slow.

But, I've been thinking that at the system taxonomies level, the `type` field (Language, Author, etc) could be implemented to determine which validation to run. That way we would only have the `SystemDefinedTaxonomy` subclass and it would not be necessary to create a subclass for each Taxonomy. 

So far we will have two subclasses: `ContentTaxonomy` and `SystemDefinedTaxonomy`

Reply:
> @pomegranited that seems fine with me in principle, as long as the code and APIs are carefully constructed to make sure `select_subclasses` is used. But this is really more of an @ormsbee question so I'd like to hear his thoughts.

I haven't used it before, but as long as the number of subclasses is relatively small, I'm fine with a few joins.

> Adding more context with this in mind. For [system-defined taxonomies](https://docs.google.com/document/d/1k2-ziCkCGT8AcGh2zorVinF0MH-2bF6_mAl0SFxr7EQ/edit#heading=h.squk6ryb7sqc), we were thinking of creating a subclass for each taxonomy, since each one had its special way of validating (`validate_object_tag`). With the current approach, in the short term it would already be slow.

What are the other types of validation that have to happen on `ObjectTag` for other taxonomy types? The reason I ask is that I think it's a little odd for the `ObjectTag` to look to their parent Taxonomy subclass to validate themselves–basically looking upward into the thing containing them. It also seems like unnecessary coupling, since we might want to tag one than one type of thing with the same Taxonomy. In fact we're effectively already doing that here with `CourseKey` and `UsageKey`.

Maybe it's ObjectTag that needs the subclassing? `CourseTag`, `BlockTag`, etc? And instead of having a completely generic `object_id` field, we make the `object_id` return a different real field value in the various subclasses (sort of like what `pk` does for Django models)? Then we could have the actual field type there–e.g. a `CourseKeyField`, instead of doing this sort of ad-hoc checking.

Reply:
> What are the other types of validation that have to happen on ObjectTag for other taxonomy types?

In the case of system-defined taxonomies, the validations that are added are more on the tag side than on the object.

- [Language](https://docs.google.com/document/d/1k2-ziCkCGT8AcGh2zorVinF0MH-2bF6_mAl0SFxr7EQ/edit#heading=h.qdyvtsaoinpj): Verify if the language is on Django LANGUAGES
- [Organization](https://docs.google.com/document/d/1k2-ziCkCGT8AcGh2zorVinF0MH-2bF6_mAl0SFxr7EQ/edit#heading=h.9rxg6ff7sw7i): Validate that the organization exists and is active.
- [Author](https://docs.google.com/document/d/1k2-ziCkCGT8AcGh2zorVinF0MH-2bF6_mAl0SFxr7EQ/edit#heading=h.r3v2eu407eyt): Validate that the User exists and is active.

Later we can add more taxonomies that have their own validation

Reply:
@ormsbee @bradenmacdonald 

> It still uses only one query, but that query does an additional JOIN per registered subclass. So whether or not this is acceptable may depend on how many subclasses there are. If it's 1-2, should be fine. 

Yeah, as @ChrisChV  noted, there's going to be a lot of these.. so using InheritanceManager like I've done in https://github.com/openedx/openedx-learning/pull/60 not a good idea.

> The reason I ask is that I think it's a little odd for the ObjectTag to look to their parent Taxonomy subclass to validate themselves–basically looking upward into the thing containing them.

I agree that `ObjectTag.is_valid` is an awkward-looking property -- I added it as a convenience method, because we're going to need the REST API to return an `is_valid` flag for object tags, so that the frontend can highlight invalid tags for the content authors to correct.

In the design stage, we pushed the ObjectTag validation into the Taxonomy class so we could avoid having to subclass both Taxonomies and ObjectTags, and link them together somehow (see below for a proposed way to do this). This made [Taxonomy.validate_object_tag()](https://github.com/openedx/openedx-learning/blob/a5b109f58dfe27c63033240c6b9458887014f517/openedx_tagging/core/tagging/models.py#L188-L219) strange, but it was intended to keep the class complexity down. But that decision was clearly flawed.

> Maybe it's ObjectTag that needs the subclassing? It could be made abstract, and have subclasses like CourseTag, BlockTag, etc?

Ok sure -- there weren't any requirements in the spec that made those particular distinctions necessary, however..

We could subclass ObjectTag to encapsulate the "closed taxonomy tags" vs "free text tags" distinction. To do this, we need a way to connect a Taxonomy with its appropriate ObjectTag subclass. To try this out, I drafted https://github.com/open-craft/openedx-learning/pull/2; see [Taxonony.object_tag_classes](https://github.com/open-craft/openedx-learning/pull/2/files#diff-204875f41064da3abfd4648611ad41c6f21abe09277263410e3daafbd343f2a3R183-R197) for an example of how this could work.

There, I also allowed for custom ObjectTag classes stored with a Taxonomy that could be used to validate system tags for the [different system taxonomy use cases listed above](https://github.com/openedx/edx-platform/pull/32518#discussion_r1251038800). We could replace the SystemTaxonomy subclasses and their validation with class methods on ObjectTag subclasses instead. Python is bad at polymorphism across class methods, but with care, we can do it, e.g

* Taxonomy subclass `LanguageTaxonomy.get_tags()` was going to return a dynamically-determined list of Tags created from the django settings.LANGUAGES.
   LanguageObjectTag could implement a class method to do the same thing.
* `AuthorTaxonomy` has too many available tags to fetch them all in a list, so it would implement `autocomplete_tags()` instead, and just return the first N dynamically-created Tags with the given prefix.
   AuthorObjectTag could implement a class method to do the same thing.

So, what do you think of something like https://github.com/open-craft/openedx-learning/pull/2 instead?

Also, if we do this, it'd be best to stop returning ObjectTag Django Models from the API altogether, and instead use the models as data to populate proper python classes and subclasses, as @bradenmacdonald [originally suggested](https://github.com/openedx/edx-platform/pull/32518#discussion_r1244335301).

If we found that we still need Taxonomy subclasses, we could add a `Taxonomy.taxonomy_class` field+property like [Taxonony.object_tag_classes](https://github.com/open-craft/openedx-learning/pull/2/files#diff-204875f41064da3abfd4648611ad41c6f21abe09277263410e3daafbd343f2a3R183-R197), and load these Taxonomy subclasses on the fly in the API using a generator, as shown with [get_object_tags()](https://github.com/open-craft/openedx-learning/pull/2/files#diff-951a5c4d32e1301ac1c30419ebce4ada8ae17db6c93f4c2a5a34caac9eb30e0aR93).

Reply:
@ormsbee  @bradenmacdonald  @ChrisChV 

Ok, this is looking really good. I'm going to close this PR in favor of https://github.com/openedx/edx-platform/pull/32661, so can we continue this conversation there?